src/Perl6/Metamodel/ParametricRoleHOW.nqp

```diff
@@ -186,7 +186,7 @@ class Perl6::Metamodel::ParametricRoleHOW
         # the concrete role.
         for self.attributes($obj, :local(1)) {
             $conc.HOW.add_attribute($conc,
-                $_.is_generic ?? $_.instantiate_generic($type_env) !! $_);
+                $_.is_generic ?? $_.instantiate_generic($type_env) !! nqp::clone($_));
```

Comment:
Hm, I think far cleaner would be for `is_generic` to return `True` when the package of the attribute is generic, and then generic instantiation to do the substitution, rather than having it happen in an ad-hoc way in `Attribute.compose`.

Reply:
Of course, that means one cannot do the other change. But attributes are never truly installed on a role, but rather exist on the classes that a role is composed in to. So it'd seem appropriate that the `package` of an attribute in a role is generic.

Reply:
@jnthn The point is that traits applying to attributes often have to consider if `$attribute.package` is valid or it is not. It's been advised a few times that `$*PACKAGE` is to be used instead of `$attribute.package` within a trait code.

The point of this PR is to reduce confusion and to have `$!package` set to what people usually expected it to be set to. I target at improving overall user experience.

With regard to `is_generic` – by what you say it would then be true for all attributes in roles until applied to a class because `$?CLASS` is generic. But that's certainly not correct.